packages/xrpc-server/src/rate-limiter.ts

```diff
@@ -0,0 +1,158 @@
+      storeClient,
+      keyPrefix: opts.keyPrefix,
+      duration: Math.floor(opts.durationMs / 1000),
+      points: opts.points,
+    })
+    return new RateLimiter(limiter, opts)
+  }
+
+  async consume(
+    ctx: XRPCReqContext,
+    opts?: { calcKey?: CalcKeyFn; calcPoints?: CalcPointsFn },
+  ): Promise<RateLimiterStatus | null> {
+    if (
+      this.byPassSecret &&
+      ctx.req.header('x-ratelimit-bypass') === this.byPassSecret
+    ) {
+      return null
+    }
+    const key = opts?.calcKey ? opts.calcKey(ctx) : this.calcKey(ctx)
+    const points = opts?.calcPoints
+      ? opts.calcPoints(ctx)
+      : this.calcPoints(ctx)
+    if (points < 1) {
+      return null
+    }
+    try {
+      const res = await this.limiter.consume(key, points)
+      return formatLimiterStatus(this.limiter, res)
+    } catch (err) {
+      // yes this library rejects with a res not an error
+      if (err instanceof RateLimiterRes) {
+        const status = formatLimiterStatus(this.limiter, err)
+        throw new RateLimitExceededError(status)
+      } else {
+        if (this.failClosed) {
+          throw err
+        }
+        logger.error(
+          {
+            err,
+            keyPrefix: this.limiter.keyPrefix,
+            points: this.limiter.points,
+            duration: this.limiter.duration,
+          },
+          'rate limiter failed to consume points',
+        )
+        return null
+      }
+    }
+  }
+}
+
+export const formatLimiterStatus = (
+  limiter: RateLimiterAbstract,
+  res: RateLimiterRes,
+): RateLimiterStatus => {
+  return {
+    limit: limiter.points,
+    duration: limiter.duration,
+    remainingPoints: res.remainingPoints,
+    msBeforeNext: res.msBeforeNext,
+    consumedPoints: res.consumedPoints,
+    isFirstInDuration: res.isFirstInDuration,
+  }
+}
+
+export const consumeMany = async (
+  ctx: XRPCReqContext,
+  fns: RateLimiterConsume[],
+): Promise<void> => {
+  if (fns.length === 0) return
+  const results = await Promise.all(fns.map((fn) => fn(ctx)))
+  const tightestLimit = getTightestLimit(results)
+  if (tightestLimit !== null) {
+    setResHeaders(ctx, tightestLimit)
+  }
+}
+
+export const setResHeaders = (
+  ctx: XRPCReqContext,
+  status: RateLimiterStatus,
+) => {
+  ctx.res.setHeader('RateLimit-Limit', status.limit)
+  ctx.res.setHeader('RateLimit-Remaining', status.remainingPoints)
+  ctx.res.setHeader(
+    'RateLimit-Reset',
+    Math.floor((Date.now() + status.msBeforeNext) / 1000),
+  )
+  ctx.res.setHeader('RateLimit-Policy', `${status.limit};w=${status.duration}`)
+}
+
+export const getTightestLimit = (
+  resps: (RateLimiterStatus | null)[],
+): RateLimiterStatus | null => {
+  let lowest: RateLimiterStatus | null = null
+  for (const resp of resps) {
+    if (resp === null) continue
+    if (lowest === null || resp.remainingPoints < lowest.remainingPoints) {
+      lowest = resp
+    }
+  }
+  return lowest
+}
+
+const defaultKey: CalcKeyFn = (ctx: XRPCReqContext) => ctx.req.ip
```

Comment:
It came-up before, but I think we may have to do something tricky to account for untrusted headers and express's handling of x-forwarded-for.

If the server knows there's one hop setting x-forwarded-for then it can take `req.ips.at(-1)`.  That should be true for us, but in general it doesn't necessarily hold.  A couple possible takeaways:
 - xrpc-server may support setting a custom default key
 - xrpc-server might expose a utility for safely determining the client ip, or bailing when it can't be determined.
 - the `calcKey()` function may support returning `null` or a specific symbol (e.g. `xrpc.kSkipRateLimit` to avoid rate-limiting when a legit key can't be determined.  That would cover our case of `req.ips.at(-1)` potentially missing.

To be honest I'm not sure there's a safe default implementation for an intermediary that appends but doesn't replace x-forwarded for 🙃 So we'll just have to choose our tradeoff and document it.  I think the ability to bail on rate limiting based on result of `calcKey` will be useful here.

Reply:
(The behavior of a missing key could also be based on whether rate limiting is configured to fail open or closed.)